File: src/host/packs.ts
// Where a language's pack comes from and where it stays.
//
// A pack is a file the core opens as bytes, not a table this side reads, so it is kept
// exactly as it arrived. It is downloaded once and held after that, because the reader who
// needs a dictionary offline is the reader on a train.
//
// The host it comes from is a runtime setting and appears nowhere in the source.
import { whole } from './whole';
import { createStore, get as read, set as keep, del, keys } from 'idb-keyval';
import { buildIpaPack, closePack, openLanguages, openPack } from '@/core';

const store = createStore('phonetix-packs', 'lang-pack');
/**
 * The packs built out of what the product carries, kept apart from the ones a reader fetched.
 *
 * Two different things under one name would be one thing: a carried pack says how a language's
 * words are said, a fetched one says what they mean, and the core holds one pack per language.
 * Separate stores keep "what this machine holds" honest - a carried pack is not a dictionary a
 * reader went and got - and let a fetched one take the open slot when it arrives.
 */
const carriedStore = createStore('phonetix-packs', 'carried-pack');

/** Languages whose open pack is one we carried, so a fetched one may take its place. */
const carried = new Set<string>();

/**
 * Where the packs are published: a release of this project, which is a static file on a host
 * somebody else keeps running.
 *
 * A pack is built once from a dump of gigabytes and does not change again, which is what a
 * release asset is for - the pronunciations the app carries are fetched the same way at build
 * time. Before this the reader was asked to name a host themselves, and since nobody served
 * them anywhere, a fresh install could fetch nothing at all.
 */
export const PUBLISHED = 'https://github.com/tieo/phonetix/releases/download/packs-v1';

/**
 * Where the packs are fetched from: where they are published, unless something has said
 * otherwise - a check serving its own, or a build pointed elsewhere.
 *
 * Read from the browser's own store rather than through the framework's helper. The helper is
 * an auto-import, and an auto-import exists only in the bundles the framework builds that way:
 * in the page the engines run in it is simply not defined, so asking for it threw, the catch
 * below turned that into "the reader has set no host", and translation was quietly off with
 * nothing anywhere saying why.
 *
 * Through `browser` rather than `chrome`, for the reason the messaging is: on Gecko the chrome
 * namespace is callback-style, so awaiting it there returned undefined and every reader on
 * Firefox was told they had set no host - no dictionaries, no translation, no explanation.
 */
export async function host(): Promise<string | undefined> {
  try {
    const got = await browser.storage.local.get('packBaseUrl');
    const said = got?.packBaseUrl;
    return typeof said === 'string' && said !== ''
      ? said.replace(/\/+$/, '')
      : PUBLISHED;
  } catch {
    // Storage being unavailable is not a reason to fail a lookup, and the published packs are
    // where they always were.
    return PUBLISHED;
  }
}

async function cached(lang: string): Promise<Uint8Array | undefined> {
  const bytes = await read<ArrayBuffer>(lang, store);
  return bytes === undefined ? undefined : new Uint8Array(bytes);
}

/** The pack built from what we carry, where one has been built already. */
async function carriedPack(lang: string): Promise<Uint8Array | undefined> {
  const bytes = await read<ArrayBuffer>(lang, carriedStore);
  return bytes === undefined ? undefined : new Uint8Array(bytes);
}

/**
 * A language's pack, open in the core, from what this machine already holds - and the
 * dictionary of what its words mean fetched in the background the first time it is needed.
 *
 * What is opened now is what is here, so a page is answered at once with how its words are
 * said; the meanings arrive a moment later and the page is told. Making the reader find a list
 * and fetch each dictionary by hand was a step nobody could work out, and until they did, a
 * reader who had chosen a language to read into was answered with nothing in it.
 */
export async function open(lang: string): Promise<string | null> {
  if ((await openLanguages()).includes(lang) && !carried.has(lang)) return lang;
  const bytes = await cached(lang);
  if (bytes) {
    carried.delete(lang);
    return openPack(bytes);
  }
  inBackground(lang);
  if ((await openLanguages()).includes(lang)) return lang;
  return fromWhatWeCarry(lang);
}

/** What is being fetched now, and what failed lately, so a language is fetched once at a time
 *  and a host that is down is not asked again on every page. */
const fetching = new Set<string>();
/** The largest dictionary fetched without being asked for. */
const BY_ITSELF_BYTES = 64 * 1024 * 1024;
const failedAt = new Map<string, number>();
const RETRY_AFTER_MS = 5 * 60_000;

function inBackground(lang: string): void {
  if (!/^[a-z]{2,3}(-[a-z0-9]+)?$/i.test(lang) || fetching.has(lang)) return;
  const failed = failedAt.get(lang);
  if (failed !== undefined && Date.now() - failed < RETRY_AFTER_MS) return;
  fetching.add(lang);
  void (async () => {
    try {
      const listed = (await offered()).find((pack) => pack.lang === lang);
      // Only what is published: asking for a pack that does not exist is a request that
      // can only fail, once per page, for every language nobody built a pack for. And only
      // what is small enough to fetch on a reader's behalf: English is every English word
      // with its English definitions, well over a hundred megabytes, and a reader of English
      // is told what a word means by their own language's dictionary, not by it.
      if (!listed || listed.bytes > BY_ITSELF_BYTES) return;
      await get(lang);
      failedAt.delete(lang);
    } catch {
      failedAt.set(lang, Date.now());
    } finally {
      fetching.delete(lang);
    }
  })();
}

/**
 * The dictionary for a language out of what this build carries.
 *
 * Every language the product knows how to pronounce ships with it, as the gzipped map it has
 * always shipped as - a fraction of the size of the pack it becomes. The pack is built the
 * first time the language is read and kept, so the cost is paid once and a reader who has
 * configured nothing at all is still answered.
 *
 * Nothing here reaches the network. What a host adds on top is what a dictionary cannot carry:
 * meanings, which are built from the dumps and are far larger.
 */
async function fromWhatWeCarry(lang: string): Promise<string | null> {
  if (!/^[a-z]{2,3}(-[a-z0-9]+)?$/i.test(lang)) return null;
  try {
    const already = await carriedPack(lang);
    if (already) {
      const opened = await openPack(already);
      carried.add(opened);
      return opened;
    }
    const res = await fetch(browser.runtime.getURL(`/dictionaries/${lang}.json.gz` as never));
    if (!res.ok) return null;
    const built = await buildIpaPack(lang, new Uint8Array(await res.arrayBuffer()));
    if (!built) return null;
    const opened = await openPack(built);
    // Kept as a pack rather than rebuilt on every start: building one is seconds for a big
    // language, and a reader opens a page more often than they install.
    await keep(opened, built.slice().buffer, carriedStore);
    carried.add(opened);
    return opened;
  } catch {
    return null;   // a build without the dictionaries answers what the host gives it
  }
}

/**
 * Fetch a language's pack, keep it, and open it.
 *
 * The pack says which language it is for and that is what comes back: a file named de.pack
 * holding Spanish would otherwise answer Spanish words to a reader who asked for German. It
 * is opened before it is kept, because a file that cannot be read is worse in the cache than
 * absent from it.
 */
export async function get(lang: string): Promise<string | null> {
  const already = await open(lang);
  // Unless what is open is one we carried: that one says how the language's words are said,
  // and what is being fetched says what they mean. The fetched pack takes the place of the
  // carried one, which is where the meanings a reader asked for come from.
  if (already && !carried.has(already)) return already;

  const base = await host();
  if (!base) return null;

  const res = await fetch(`${base}/${lang}.pack`);
  if (!res.ok) throw new Error(`${res.status} fetching the ${lang} pack`);
  const bytes = await whole(res);
  // Checked against the listing where it says what the file has to be: a pack cut short opens
  // as a pack that is missing words, which is worse than not having it.
  const listed = (await offered().catch(() => [])).find((pack) => pack.lang === lang);
  if (listed?.sha256) {
    const digest = await crypto.subtle.digest('SHA-256', bytes);
    const got = [...new Uint8Array(digest)].map((b) => b.toString(16).padStart(2, '0')).join('');
    if (got !== listed.sha256) throw new Error(`the ${lang} pack arrived as ${got}`);
  }
  const opened = await openPack(bytes);
  await keep(opened, bytes.slice().buffer, store);
  carried.delete(opened);
  await told();
  return opened;
}

/**
 * Say which dictionaries this machine holds, where a page can hear it.
 *
 * A page reading itself has no way of knowing a dictionary has arrived: it redraws when a
 * setting changes, and which packs are held is not a setting. Without this, a reader who
 * fetched the dictionary for the page in front of them sat looking at the same bare page until
 * they touched something else or loaded it again.
 */
async function told(): Promise<void> {
  try {
    await browser.storage.local.set({ heldPacks: await held() });
  } catch {
    // A page that cannot be told redraws on the next setting it is given, as it did before.
  }
}

/** One pack as the host that serves them describes it, which is what packbuild writes. */
export interface Offered {
  id: string;
  lang: string;
  built: number;
  entries: number;
  keys: number;
  glosses: number;
  bytes: number;
  sha256: string;
}

/** The listing, asked for once a while rather than once per question. */
let listing: { base: string; at: number; packs: Offered[] } | null = null;
const LISTING_FOR_MS = 10 * 60_000;

/** What the published packs are, or what a host of the reader's own offers instead. */
export async function offered(): Promise<Offered[]> {
  const base = await host();
  if (!base) return [];
  if (listing && listing.base === base && Date.now() - listing.at < LISTING_FOR_MS) {
    return listing.packs;
  }
  const res = await fetch(`${base}/packs.json`);
  if (!res.ok) throw new Error(`${res.status} fetching the list of packs`);
  const listed = (await res.json()) as Offered[] | { packs: Offered[] };
  const packs = Array.isArray(listed) ? listed : (listed.packs ?? []);
  listing = { base, at: Date.now(), packs };
  return packs;
}

/** Which languages are held on this machine, open or not. */
export async function held(): Promise<string[]> {
  return (await keys(store)).map(String).sort();
}

/** Give up a pack, on the machine and in the core. */
export async function forget(lang: string): Promise<void> {
  await del(lang, store);
  await closePack(lang);
  await told();
  // And back to what the product carries, so giving up a dictionary leaves the language
  // readable rather than bare.
  await fromWhatWeCarry(lang);
}

File: android/app/src/main/java/io/github/tieo/phonetix/core/Packs.kt
package io.github.tieo.phonetix.core

import android.content.Context
import org.json.JSONArray
import java.io.File
import java.net.URL

/** One pack as the host that serves them describes it, which is what packbuild writes. */
data class Offered(
    val lang: String,
    val entries: Int,
    val bytes: Long,
    /** What the file has to hash to, where the listing says. */
    val sha256: String = "",
    /** Where it is served from: the reader's own host, or the published release for what that
     *  host does not have. */
    val at: String = "",
)

/**
 * The dictionaries this phone has, and the ones it could have.
 *
 * A pack is a file the core opens where it lies, so it is downloaded once into the app's own
 * files and kept. Nothing is fetched because a screen happened to be in a language: a
 * dictionary is tens of megabytes and the reader decides which ones they want.
 *
 * Where they come from is a runtime setting and appears nowhere in the source.
 */
object Packs {

    /** Which languages are open in the core right now. Written by the thread that reads the
     *  screen and by the one dictionaries arrive on, so a set that can be. */
    private val open: MutableSet<String> = java.util.concurrent.ConcurrentHashMap.newKeySet()

    private fun file(context: Context, lang: String) =
        File(context.applicationContext.filesDir, "lex-$lang.pack")

    /** Which dictionaries this phone holds. */
    fun held(context: Context): List<String> =
        context.applicationContext.filesDir.listFiles()
            ?.mapNotNull { file ->
                file.name.takeIf { it.startsWith("lex-") && it.endsWith(".pack") }
                    ?.removePrefix("lex-")?.removeSuffix(".pack")
            }
            ?.sorted()
            ?: emptyList()

    /** Open everything this phone holds, so the cascade can answer with it. */
    fun openHeld(context: Context) {
        if (Reading.core == 0L) return
        for (lang in held(context)) {
            if (lang in open) continue
            val opened = runCatching { Lex.openPack(Reading.core, file(context, lang).path) }
                .getOrNull()
            if (!opened.isNullOrEmpty()) {
                open.add(opened)
                openClassifier(context, opened)
            }
        }
    }

    /** Which languages carry a trained homograph classifier, so none is looked for in vain. */
    private val trained = setOf("de", "en", "es", "fr", "it", "ja", "nl", "pt", "ru", "zh")
    private val classified: MutableSet<String> = java.util.concurrent.ConcurrentHashMap.newKeySet()

    /**
     * Give the core a language's homograph classifier, once.
     *
     * Unpacked from the apk on first use, since the engine reads it as a file. A language with
     * none loses nothing: the word before it still decides where it decides.
     */
    fun openClassifier(context: Context, lang: String) {
        // Claimed atomically, so two threads meeting the same language do not both unpack it.
        if (lang !in trained || Reading.core == 0L || !classified.add(lang)) return
        val into = File(context.applicationContext.filesDir, "homographs")
        val path = File(into, "$lang.hg")
        runCatching {
            if (!path.exists()) {
                into.mkdirs()
                context.assets.open("homographs/$lang.hg").use { source ->
                    path.outputStream().use { source.copyTo(it) }
                }
            }
            val known = Lex.openHomographs(Reading.core, lang, path.absolutePath)
            if (io.github.tieo.phonetix.BuildConfig.DEBUG) {
                android.util.Log.d("Phonetix", "HOMOGRAPHS $lang knows $known")
            }
        }.onFailure { android.util.Log.w("Phonetix", "no classifier for $lang", it) }
    }

    /**
     * Where the packs are published: a release of this project, which is a static file on a
     * host somebody else keeps running.
     *
     * A pack is built once out of a dump of gigabytes and does not change again, which is
     * what a release asset is for - the pronunciations the app carries are fetched the same
     * way when it is built. Before this the reader was asked to name a host themselves, and
     * since nobody served them anywhere, a fresh install could fetch nothing at all.
     */
    const val PUBLISHED = "https://github.com/tieo/phonetix/releases/download/packs-v1"

    /** Where to fetch from: what was asked for, or where they are published. */
    private fun from(base: String): String =
        base.ifBlank { PUBLISHED }.trimEnd('/')

    /**
     * What there is to be had, and where each comes from.
     *
     * A host of the reader's own is asked first, and whatever it does not have comes from the
     * published release. A host set up before the packs were published serves an older set -
     * or has gone - and a phone that asked it alone went on with neither the dictionaries it
     * lacked nor any saying so.
     */
    fun offered(base: String): List<Offered> {
        val theirs = if (base.isBlank()) emptyList() else listed(from(base))
        val published = listed(PUBLISHED)
        return theirs + published.filter { mine -> theirs.none { it.lang == mine.lang } }
    }

    private fun listed(host: String): List<Offered> {
        listings[host]?.let { (at, packs) ->
            if (android.os.SystemClock.elapsedRealtime() - at < LISTING_FOR_MS) return packs
        }
        return fetchListing(host).also {
            if (it.isNotEmpty()) listings[host] = android.os.SystemClock.elapsedRealtime() to it
        }
    }

    /** Each host's listing, asked for once a while rather than once per dictionary. */
    private val listings = java.util.concurrent.ConcurrentHashMap<String, Pair<Long, List<Offered>>>()
    private const val LISTING_FOR_MS = 10 * 60_000L

    private fun fetchListing(host: String): List<Offered> {
        val text = runCatching { fetchText("$host/packs.json") }
            .getOrNull() ?: return emptyList()
        val listed = runCatching { JSONArray(text) }.getOrNull() ?: return emptyList()
        return (0 until listed.length()).mapNotNull { at ->
            listed.optJSONObject(at)?.let { row ->
                val lang = row.optString("lang")
                if (lang.isEmpty()) {
                    null
                } else {
                    Offered(
                        lang, row.optInt("entries"), row.optLong("bytes"), row.optString("sha256"),
                        host,
                    )
                }
            }
        }
    }

    /**
     * Fetch a language's dictionary and open it.
     *
     * Opened before it is kept: a file that cannot be read is worse in the cache than absent
     * from it, and the pack says which language it turned out to be for.
     */
    fun get(context: Context, base: String, lang: String): Boolean {
        if (Reading.core == 0L) return false
        // Only what the listing offers, and checked against what it says the file hashes to:
        // asking for a language nobody built a pack for is a request that can only fail, and a
        // pack cut short opens as a dictionary quietly missing words.
        val listed = offered(base).firstOrNull { it.lang == lang } ?: return false
        val into = file(context, lang)
        val arriving = File(into.parentFile, "${into.name}.new")
        if (!download("${listed.at.ifEmpty { from(base) }}/$lang.pack", arriving, listed.sha256)) {
            return false
        }
        return runCatching {
            val opened = Lex.openPack(Reading.core, arriving.path)
            require(opened.isNotEmpty()) { "not a pack" }
            check(arriving.renameTo(into)) { "could not keep the $lang pack" }
            open.add(opened)
            true
        }.onFailure {
            arriving.delete()
            android.util.Log.w("Phonetix", "the $lang pack did not open", it)
        }.getOrDefault(false)
    }

    /** Where the translation models are kept, one directory per direction. */
    fun models(context: Context): File =
        File(context.applicationContext.filesDir, "models")

    /**
     * Fetch the model for one direction, from the same place the packs come from.
     *
     * Three or four files and tens of megabytes, fetched once. The listing names each file,
     * and where it names a full address and a checksum, that is where the file comes from and
     * what it has to be: the published listing points at the models where their makers
     * publish them, pinned to the versions this app was checked against, rather than copying
     * them somewhere else. A host of the reader's own may list bare names instead, which are
     * then looked for beside the listing.
     */
    fun getModel(context: Context, base: String, from: String, to: String): Boolean =
        oneModel(context, base, from, to) ||
            Translator.through(from, to).let { steps ->
                // Through English where the pair has no model of its own; see
                // [Translator.through].
                steps.isNotEmpty() && steps.all { (a, b) ->
                    Translator.ready(models(context), a, b) || oneModel(context, base, a, b)
                }
            }

    private fun oneModel(context: Context, base: String, from: String, to: String): Boolean =
        modelFrom(context, from(base), from, to) ||
            (base.isNotBlank() && modelFrom(context, PUBLISHED, from, to))

    /** The model for one direction from one host, where that host lists it. */
    private fun modelFrom(context: Context, host: String, from: String, to: String): Boolean {
        val listed = runCatching { JSONArray(fetchText("$host/models.json")) }
            .onFailure { android.util.Log.w("Phonetix", "no model listing at the host", it) }
            .getOrNull() ?: return false
        val row = (0 until listed.length())
            .mapNotNull { listed.optJSONObject(it) }
            .firstOrNull { it.optString("from") == from && it.optString("to") == to }
            ?: return false
        val files = row.optJSONObject("files") ?: return false
        val into = File(models(context), "$from-$to")
        into.mkdirs()
        // Under the names they are published with. The engine reads the kind of model it is
        // from the file name - a model built for intgemm says so there - so a file renamed to
        // something tidier is a model it cannot use.
        for (key in MODEL_FILES) {
            val file = files.optJSONObject(key)
            val named = file?.optString("name").orEmpty()
            if (named.isEmpty()) {
                // A model and a vocabulary are what a direction is; the shortlist, and a
                // vocabulary split in two for a pair whose scripts differ, are optional.
                if (key == "model") return false
                continue
            }
            val kept = File(into, named)
            val wanted = file?.optString("sha256").orEmpty()
            if (kept.exists() && kept.length() > 0 && (wanted.isEmpty() || sha256(kept) == wanted)) {
                continue
            }
            val address = file?.optString("url").orEmpty().ifEmpty { "$host/models/$named" }
            if (!download(address, kept, wanted)) return false
        }
        return true
    }

    /** The files a direction can be made of, in the order they are fetched. */
    private val MODEL_FILES = listOf("model", "vocab", "srcvocab", "trgvocab", "lex")

    /**
     * Fetch a file to where it is kept, whole or not at all.
     *
     * Into a partial file first and renamed only once complete and, where a checksum is
     * known, only once it matches: a download that stopped halfway, or a host that answered
     * with an error page, is otherwise a file the engine is handed as a model.
     */
    fun download(address: String, to: File, sha256: String = ""): Boolean {
        val part = File(to.parentFile, "${to.name}.part")
        return runCatching {
            open(address).use { source -> part.outputStream().use { sink -> source.copyTo(sink) } }
            if (sha256.isNotEmpty()) {
                val got = sha256(part)
                check(got == sha256) { "$address arrived as $got, not $sha256" }
            }
            check(part.renameTo(to)) { "could not keep $to" }
        }.onFailure {
            part.delete()
            android.util.Log.w("Phonetix", "$address did not arrive", it)
        }.isSuccess
    }

    /** A connection that gives up rather than hanging: a stalled host is a panel waiting. */
    private fun open(address: String): java.io.InputStream {
        val connection = URL(address).openConnection() as java.net.HttpURLConnection
        connection.connectTimeout = CONNECT_MS
        connection.readTimeout = READ_MS
        connection.instanceFollowRedirects = true
        val code = connection.responseCode
        check(code in 200..299) { "$address answered $code" }
        return connection.inputStream
    }

    private fun fetchText(address: String): String =
        open(address).use { it.readBytes().toString(Charsets.UTF_8) }

    private fun sha256(file: File): String {
        val digest = java.security.MessageDigest.getInstance("SHA-256")
        file.inputStream().use { source ->
            val buffer = ByteArray(1 shl 16)
            while (true) {
                val read = source.read(buffer)
                if (read < 0) break
                digest.update(buffer, 0, read)
            }
        }
        return digest.digest().joinToString("") { "%02x".format(it) }
    }

    private const val CONNECT_MS = 15_000
    private const val READ_MS = 30_000

    /** Give a dictionary up. */
    fun forget(context: Context, lang: String) {
        file(context, lang).delete()
        open.remove(lang)
        if (Reading.core != 0L) runCatching { Lex.closePack(Reading.core, lang) }
        // And the pronunciations this app carries answer that language again: they were held
        // back while a fetched dictionary was there, since only one pack per language answers.
        Dictionary.released(lang)
        Dictionary.ensure(context, lang)
    }
}
